Hid joystick interface (#4226)
qmk/qmk_firmware: docs/feature_joystick.md

## Joystick HID Device

The keyboard can be made to be recognized as a joystick HID device by the Operating System. 

This is enabled by adding the following to `rules.mk`

```
JOYSTICK_ENABLE = yes
```

The joystick feature provides two services : 
 * reading an analog input device
 * sending gamepad HID reports

Both services can be used without the other, depending on wether you just want to read a device but not send gamepad reports (for volume control for instance) 
or send gamepad reports based on values computed by the keyboard.

### Analog circuit

> An analog device such as a potentiometer found on a gamepad's analog axes is based on a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider).

Author: Agreed

## Joystick HID Device

The keyboard can be made to be recognized as a joystick HID device by the Operating System.

This is enabled by adding the following to `rules.mk`

```makefile
JOYSTICK_ENABLE = yes
```

!> Joystick support is not currently available on V-USB devices.

The joystick feature provides two services :
 * reading an analog input device
 * sending gamepad HID reports

Both services can be used without the other, depending on whether you just want to read a device but not send gamepad reports (for volume control for instance)
or send gamepad reports based on values computed by the keyboard.

> !> Reading analog axes is not currently functional on ARM MCUs.

Author: This is tested on a proton c already. You can try it for confirmation if you'd like. With the onekey firmware, the first axis is read from A0.

### Analog circuit

An analog device such as a potentiometer found on a gamepad's analog axes is based on a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider).
It is composed of three connectors linked to the ground, the power input and power output (usually the middle one). The power output holds the voltage that varies based on the position of the cursor,
which value will be read using your MCU's [ADC](https://en.wikipedia.org/wiki/Analog-to-digital_converter).
Depending on what pins are already used by your keyboard's matrix, the rest of the circuit can get a little bit more complicated,
feeding the power input and ground connection through pins and using diodes to avoid bad interactions with the matrix scanning procedures.

### Configuring the joystick

The default joystick has 2 axes and and 8 buttons. This can be changed from the config.h file :

> 

Author: Yeah, it seemed like a sensible default to have axes and buttons to match a usual joystick.

## Joystick HID Device

The keyboard can be made to be recognized as a joystick HID device by the Operating System. 

> This is enable by adding in your rule.mk : 

Reviewer: Just some wordsmithing

```suggestion
This is enabled by adding the following to `rules.mk`
```

//max 6 for JOYSTICK_AXES_COUNT
#define JOYSTICK_AXES_COUNT 3
```

When defining axes for your joystick, you have to provide a definition array for it. You can do this from your keymap.c file.
A joystick will either be read from an input pin that allows the use of an ADC, or can be virtual, so that its value is provided by your code.
You have to define an array of type ''joystick_config_t'' and of proper size, in the following way : 

```
//joystick config
joystick_config_t joystick_axes[JOYSTICK_AXES_COUNT] = {
  [0] = {D7, 65280, 65472},
  [1] = {JS_VIRTUAL_AXIS, 65280, 65472}
};
```

In this example, the first axis will be read from the D7 pin, using an analogRead, whereas the second axis will not be read.
In order to give a value to the second axis, you can do so in any custimizable entry point of quantum : as an action, in process_record_user or in matrix_scan_user, or even in joystick_task(void) which is call even when no key has been pressed.
You assign a value by writing to joystick_status.axes[axis_index] a signed 8bit value (-127 to 127). Then it is necessary to assign the flag JS_UPDATED to joystick_status.status in order for the change to be taken into account.

> 

Reviewer 2: Would like to have some clarification here. Is the `joystick_config_t` array necessary if you only want to use virtual joystick axis like described below, and if it is necessary, how to declare it since there is no actual pin with analog data?

Author: Yes, the joystick_config_t is still necessary. I might be able to make it optional if all axes are virtual though. For now, the config would be as described in the doc, with all pin definitions defined as JS_VIRTUAL_AXIS. The two values after define the range of analog value read to map to -127;127, so there is no use for it for virtual axes.

Author: @coredump I saw your post on reddit and that you are using a pro micro. Have you tried building a firmware with the joystick feature enabled yet?

Reviewer 2: I did, it compiled, but I didn't add axis, just buttons. I got an error on Windows, like USB device not recognized. I saw some people on the Lufa github saying something about avoiding declaring a serial for the device (there's a env variable for that), but I haven't tested it yet.

Reviewer 2: So, for some reason this is not compiling anymore. It's complaining about not finding the `joystick_config_t ` or anything defined on the quantum/* files related to the joystick support. I tried to find out why, but afaik the rules.mk is correct, and the common_features.mk should be adding the files to the compilation, but it's not happening.
There's a gist here with the error: https://gist.github.com/coredump/83b08aaab5e194f74785c2b94dd8c38d

(also I am running from your branch, @a-chol )

Reviewer 2: I am doing the definition like this
```c
#ifdef JOYSTICK_ENABLE
joystick_config_t joystick_axes[JOYSTICK_AXES_COUNT] = {
  [0] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [1] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [2] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [3] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [4] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [5] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [6] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [7] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
};
#endif
```

The issue is that the files aren't being found or used by the compiler 

```console
Compiling: keyboards/fightkeyb/keymaps/default/keymap.c  
    keyboards/fightkeyb/keymaps/default/keymap.c:31:1: error: unknown type name 'joystick_config_t'; did you mean 'keymap_config_t'?
 joystick_config_t joystick_axes[JOYSTICK_AXES_COUNT] = {```

are you able to do a `make` successfully on your branch? I don't know why mine would fail.

Author: Did you
#include <quantum/ joystick.h>
in your keymap.c?

Reviewer 2: Ah. That was it.  is that on the doc somewhere?

Now I am able to compile and upload the firmware. My linux kernel is full of this:

```Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0```

And I can't see a joystick yet, but I will continue digging.

Author: You're right, that's an oversight on my part, I'll add it. I'll check some more in the evening (CET) to see if I spot any mistake in the Lufa implementation. Do you have a tool that could display the hid report descriptors received for the device?

Reviewer 2: https://gist.github.com/coredump/e924fb59d2b9301cf8e7f18b128641a1 I got some HID dumping done. I changed the `PRODUCT_ID` and `DEVICE_VERSION` on my config.h because it was conflicting with my other QMK keyboard on the same machine.

Author: I spotted a few mistakes that could provide an incorrect device descriptor. Can you try to apply the following patch and see what changes?
https://gist.github.com/a-chol/97cbf2e1e1d7952324663459b817072c

Author: Alright, that looks better. The only thing that seems off is the 
Report Count (7),
for the joysticks. Only 6 axes are supported, six usages defined just before so there is no seventh axis. Did you define JOYSTICK_AXES_COUNT 7 or 6?
Regarding the buttons, it looks good. There are 9 usages, 9 1-bit reports. You say that it reports the wrong number of buttons, did you not define JOYSTICK_BUTTON_COUNT to 9? What value did you give it?
Do the errors appear synchronously with you hitting buttons mapped to JS_BUTTON* keycodes?

Reviewer 2: So, last message, I meant to say that it was reporting the wrong number of axis, not buttons, buttons are reported fine. I tried doing some changes to the report record to match some joysticks/gamepads I have here, so I ended with https://gist.github.com/coredump/f4cb72b5d6a0d2595d359527de0b0e37

I had the wrong number of axis on my config (I was using 7, sorry about that, off-by-1). I changed it to the default of 2 axis for testing now.

After that HID change, jstest started to show the correct number of axis(2).

Unfortunately, the USB-reset errors still happening and no input is seen by `evtest` or `jstest`. The resets don't happen in sync with the button press, it just keeps resetting as long as the testing programs are running, my guess is that the programs are trying to open the device and keep failing (that would explain the reset and the fact that I can't get any event coming from the joystick'. I don't think 

Interesting enough, since we have 2 report records (one for a keyboard, one for a joystick), I get two inputs on my kernel, and the keyboard input works fine (I have some buttons sending normal KC keycodes), it's just the joystick input that doesn't open/connect.

Author: can you try applying the following patch?
https://gist.github.com/a-chol/3bdf405a65500c335148372edc8860a2
The send_joystick_packet function seems fine otherwise.

Reviewer 2: That worked! At least on linux.
I got it to work and report the events to my js testing stuff. I rolled back all my changes to the HID descriptor too to check, and it seems to work, with one little caveat: if I use only 2 axis (0 and 1), this happens:
evtest = sees 2 axis
jstest = sees 0 axis
sdl2-jstest = sees 9 buttons and a ball (trackball)

So I added 6 axis back, and all of them report as axis then, for all the testing programs (so, linux input event subsystem and the joydev joystick driver).

Unfortunately, it appears as a broken USB device on windows, something about not being able to start the device, error `0xA` and `0xC000009C` that I searched and didn't get too far, as everything related to windows error codes. I did get some better error from a USB tool on widows and got `Report Descriptor was not byte aligned.` that sounds more likely to be a fixable bug.

But it is looking great!

Author: Good new for linux! Can you provide me the diff of the changes you made, or offer a PR on my fork (if that's a thing, I'm not well versed in the usages of github). 
I use USBlyzer on windows, did you use that or does it give any more info. I don't really see how the report descriptor could be misaligned. 
Maybe you could try returning early from send_joystick_packet in lufa.c to avoid any report sending ; to check that it's not the report that is creating the error.
You could also vary the number of buttons : I suspect that using a number of buttons that is not a multiple of 8 wouldn't be handled well. I think I send the whole eight-bit array when I should only be sending the relevant bits. I'll look into it tomorrow.

Reviewer 2: I forgot to let you know that I added a PR.

Author: Excellent, it's now merged into this one! I'll add your fix for byte-aligned reports to the other existing implementations. Thanks for your help, hope you enjoy the feature!

Reviewer 3: This got resolved, and added in, correct?

Author: Yes, this got merged in my branch.

Reviewer 3: Just adding some feedback here, it looks like you still need to manually add the include.  That's .... not good practice, and should be automatically included in "quantum.h" if the feature is enabled.
